File: products/chargeback/ui/src/lib/access.test.ts
import { describe, expect, it } from 'vitest'
import type { Me } from '../api/types'
import { can, canPartner, customerIds, customerScope, displayRole, homeFor, isPartner, isSovereign, partnerIds, partnerName, primaryPartnerId, roleLabel, scopeKindOf } from './access'

const A = '11111111-1111-1111-1111-111111111111'
const B = '22222222-2222-2222-2222-222222222222'

const sovereignAdmin: Me = {
  email: 'ops@nc.example',
  role: 'operator',
  roles: [{ role: 'sovereign-admin', scope_kind: 'sovereign', source: 'config' }],
  permissions: { sovereign: ['metering.read', 'rating.manage', 'customers.manage', 'billing.issue', 'billing.collect', 'account.topup', 'settings.manage', 'audit.read', 'customer.self.manage'] },
  scopes: ['sovereign'],
}
const financeViewer: Me = {
  email: 'fin@nc.example',
  role: 'finance-viewer',
  roles: [{ role: 'finance-viewer', scope_kind: 'sovereign', source: 'group:finance' }],
  permissions: { sovereign: ['audit.read', 'metering.read'] },
  scopes: ['sovereign'],
}
const owner: Me = {
  email: 'owner@acme.example',
  role: 'customer-admin',
  customer_id: A,
  roles: [{ role: 'customer-owner', scope_kind: 'customer', customer_id: A, source: 'binding' }],
  permissions: { [customerScope(A)]: ['account.topup', 'customer.self.manage', 'metering.read'] },
  scopes: [customerScope(A)],
}
const viewer: Me = {
  email: 'v@acme.example',
  role: 'customer-viewer',
  customer_id: A,
  roles: [{ role: 'customer-viewer', scope_kind: 'customer', customer_id: A, source: 'binding' }],
  permissions: { [customerScope(A)]: ['metering.read'] },
  scopes: [customerScope(A)],
}

describe('access — the console decides by permissions', () => {
  it('a Sovereign permission covers every customer', () => {
    expect(isSovereign(sovereignAdmin)).toBe(true)
    expect(can(sovereignAdmin, 'settings.manage')).toBe(true)
    expect(can(sovereignAdmin, 'billing.collect', A)).toBe(true)
    expect(can(sovereignAdmin, 'customer.self.manage', B)).toBe(true)
    expect(homeFor(sovereignAdmin)).toBe('/overview')
  })

  it('a finance-viewer reads and changes nothing', () => {
    expect(isSovereign(financeViewer)).toBe(true)
    expect(can(financeViewer, 'metering.read', A)).toBe(true)
    expect(can(financeViewer, 'billing.issue')).toBe(false)
    expect(can(financeViewer, 'settings.manage')).toBe(false)
    expect(can(financeViewer, 'rating.manage')).toBe(false)
    expect(displayRole(financeViewer)).toBe('finance-viewer')
    expect(roleLabel(displayRole(financeViewer))).toBe('Finance viewer')
  })

  it('a customer-owner holds its permissions on its own customer only', () => {
    expect(isSovereign(owner)).toBe(false)
    expect(can(owner, 'account.topup', A)).toBe(true)
    expect(can(owner, 'customer.self.manage', A)).toBe(true)
    expect(can(owner, 'metering.read', A)).toBe(true)
    expect(can(owner, 'account.topup', B)).toBe(false)
    expect(can(owner, 'account.topup')).toBe(false)
    expect(can(owner, 'billing.collect', A)).toBe(false)
    expect(can(owner, 'billing.issue', A)).toBe(false)
    expect(customerIds(owner)).toEqual([A])
    expect(homeFor(owner)).toBe('/my/overview')
    expect(roleLabel(displayRole(owner))).toBe('Owner')
  })

  it('a customer-viewer cannot top up or manage users', () => {
    expect(can(viewer, 'metering.read', A)).toBe(true)
    expect(can(viewer, 'account.topup', A)).toBe(false)
    expect(can(viewer, 'customer.self.manage', A)).toBe(false)
  })

  it('the most powerful binding names the principal', () => {
    const both: Me = {
      ...owner,
      roles: [
        { role: 'customer-owner', scope_kind: 'customer', customer_id: A },
        { role: 'finance-viewer', scope_kind: 'sovereign' },
      ],
      permissions: { ...owner.permissions, sovereign: ['metering.read', 'audit.read'] },
      scopes: ['sovereign', customerScope(A)],
    }
    expect(displayRole(both)).toBe('finance-viewer')
    expect(isSovereign(both)).toBe(true)
    expect(customerIds(both)).toEqual([A])
  })

  it('a /me document from before bindings still decides by the legacy role', () => {
    const oldOperator: Me = { email: 'ops@nc.example', role: 'operator' }
    const oldAdmin: Me = { email: 'adm@acme.example', role: 'customer-admin', customer_id: A }
    const oldViewer: Me = { email: 'v@acme.example', role: 'customer-viewer', customer_id: A }
    expect(isSovereign(oldOperator)).toBe(true)
    expect(can(oldOperator, 'settings.manage')).toBe(true)
    expect(can(oldAdmin, 'customer.self.manage', A)).toBe(true)
    expect(can(oldAdmin, 'billing.collect', A)).toBe(false)
    expect(can(oldAdmin, 'metering.read', B)).toBe(false)
    expect(can(oldViewer, 'account.topup', A)).toBe(false)
    expect(customerIds(oldAdmin)).toEqual([A])
    expect(homeFor(oldAdmin)).toBe('/my/overview')
  })

  it('nobody signed in holds nothing', () => {
    expect(can(null, 'metering.read')).toBe(false)
    expect(isSovereign(null)).toBe(false)
    expect(customerIds(null)).toEqual([])
    expect(homeFor(null)).toBe('/signin')
  })

  it('every role has one scope kind', () => {
    expect(scopeKindOf('sovereign-admin')).toBe('sovereign')
    expect(scopeKindOf('billing-operator')).toBe('sovereign')
    expect(scopeKindOf('finance-viewer')).toBe('sovereign')
    expect(scopeKindOf('partner-owner')).toBe('partner')
    expect(scopeKindOf('partner-viewer')).toBe('partner')
    expect(scopeKindOf('customer-owner')).toBe('customer')
    expect(scopeKindOf('customer-billing')).toBe('customer')
    expect(scopeKindOf('customer-viewer')).toBe('customer')
    expect(scopeKindOf('root')).toBe('')
  })
})

// The partner scope (DESIGN.md §11.5): a binding at partner:<id> covers the
// customers it expanded to, and its partner-scoped permissions are asked for
// at the partner, never at the Sovereign.
const P = '99999999-9999-9999-9999-999999999999'
const partnerOwner: Me = {
  email: 'ap@resell.example',
  role: 'partner-owner',
  customer_id: 'party-1',
  roles: [{ role: 'partner-owner', scope_kind: 'partner', partner_id: P, partner_name: 'Resell Co', customer_ids: ['party-1', A] }],
  permissions: { [`partner:${P}`]: ['metering.read', 'account.topup', 'partner.self.manage'] },
  scopes: [`partner:${P}`],
}

describe('the partner lens', () => {
  it('is a lens of its own, neither the Sovereign nor a customer', () => {
    expect(isPartner(partnerOwner)).toBe(true)
    expect(isSovereign(partnerOwner)).toBe(false)
    expect(isPartner(sovereignAdmin)).toBe(false)
    expect(homeFor(partnerOwner)).toBe('/partner/overview')
    expect(partnerIds(partnerOwner)).toEqual([P])
    expect(primaryPartnerId(partnerOwner)).toBe(P)
    expect(partnerName(partnerOwner, P)).toBe('Resell Co')
  })

  it('covers the customers it expanded to, and no others', () => {
    expect(can(partnerOwner, 'metering.read', A)).toBe(true)
    expect(can(partnerOwner, 'metering.read', B)).toBe(false)
    expect(customerIds(partnerOwner)).toContain(A)
    expect(customerIds(partnerOwner)).toContain('party-1')
    // A partner-scoped permission is not a Sovereign one.
    expect(can(partnerOwner, 'metering.read')).toBe(false)
    expect(can(partnerOwner, 'partners.manage', A)).toBe(false)
  })

  it('asks partner permissions at the partner', () => {
    expect(canPartner(partnerOwner, 'partner.self.manage', P)).toBe(true)
    expect(canPartner(partnerOwner, 'partner.self.manage', 'another')).toBe(false)
    expect(canPartner(partnerOwner, 'partners.manage', P)).toBe(false)
    // A Sovereign binding answers at every partner.
    expect(canPartner(sovereignAdmin, 'metering.read', P)).toBe(true)
    expect(canPartner(null, 'metering.read', P)).toBe(false)
  })

  it('labels the two partner roles', () => {
    expect(roleLabel('partner-owner')).toBe('Partner owner')
    expect(roleLabel('partner-viewer')).toBe('Partner viewer')
    expect(displayRole(partnerOwner)).toBe('partner-owner')
  })
})

File: products/chargeback/internal/access/access_test.go
package access

import (
	"reflect"
	"testing"

	"github.com/openova-io/openova/products/chargeback/internal/store"
)

func ptr(s string) *string { return &s }

// TestMatrixEveryRoleEveryPermission pins the whole policy: every role ×
// every permission, at both scope kinds. A change to Matrix must change this
// table on purpose.
func TestMatrixEveryRoleEveryPermission(t *testing.T) {
	want := map[string]map[Permission]bool{
		RoleSovereignAdmin:  {MeteringRead: true, RatingManage: true, CustomersManage: true, BillingIssue: true, BillingCollect: true, AccountTopup: true, SettingsManage: true, AuditRead: true, CustomerSelfManage: true, PartnersManage: true, PartnerSelfManage: true},
		RoleBillingOperator: {MeteringRead: true, RatingManage: true, CustomersManage: true, BillingIssue: true, BillingCollect: true, AuditRead: true, CustomerSelfManage: true, PartnersManage: true, PartnerSelfManage: true},
		RoleFinanceViewer:   {MeteringRead: true, AuditRead: true},
		RolePartnerOwner:    {MeteringRead: true, AccountTopup: true, PartnerSelfManage: true},
		RolePartnerViewer:   {MeteringRead: true},
		RoleCustomerOwner:   {MeteringRead: true, AccountTopup: true, CustomerSelfManage: true},
		RoleCustomerBilling: {MeteringRead: true, AccountTopup: true},
		RoleCustomerViewer:  {MeteringRead: true},
	}
	if len(want) != len(store.Roles) {
		t.Fatalf("table covers %d roles, store.Roles has %d", len(want), len(store.Roles))
	}
	for _, role := range store.Roles {
		for _, perm := range Permissions {
			if got := RoleGrants(role, perm); got != want[role][perm] {
				t.Errorf("%s × %s = %v, want %v", role, perm, got, want[role][perm])
			}
		}
	}
}

// TestScopesSovereignCoversEveryCustomerCustomerCoversOnlyItsOwn is the scope
// rule: a Sovereign binding answers at the Sovereign and on any customer; a
// customer binding answers on its customer only and never at the Sovereign.
func TestScopesSovereignCoversEveryCustomerCustomerCoversOnlyItsOwn(t *testing.T) {
	a, b := "aaaaaaaa-0000-0000-0000-000000000001", "bbbbbbbb-0000-0000-0000-000000000002"
	for _, role := range []string{RoleSovereignAdmin, RoleBillingOperator, RoleFinanceViewer} {
		bs := []store.RoleBinding{{Role: role, ScopeKind: ScopeSovereign}}
		for _, perm := range Matrix[role] {
			if !Has(bs, perm, "") || !Has(bs, perm, a) || !Has(bs, perm, b) {
				t.Errorf("%s: %s must hold at the Sovereign and on every customer", role, perm)
			}
		}
	}
	for _, role := range []string{RoleCustomerOwner, RoleCustomerBilling, RoleCustomerViewer} {
		bs := []store.RoleBinding{{Role: role, ScopeKind: ScopeCustomer, CustomerID: ptr(a)}}
		for _, perm := range Matrix[role] {
			if !Has(bs, perm, a) {
				t.Errorf("%s: %s must hold on its own customer", role, perm)
			}
			if Has(bs, perm, b) {
				t.Errorf("%s: %s must not hold on another customer", role, perm)
			}
			if Has(bs, perm, "") {
				t.Errorf("%s: %s must not hold at the Sovereign", role, perm)
			}
		}
	}
}

// TestDiscriminatingCasePerRole names the one decision that tells each role
// from its neighbours.
func TestDiscriminatingCasePerRole(t *testing.T) {
	a, b := "aaaaaaaa-0000-0000-0000-000000000001", "bbbbbbbb-0000-0000-0000-000000000002"
	p := "pppppppp-0000-0000-0000-000000000003"
	sov := func(role string) []store.RoleBinding {
		return []store.RoleBinding{{Role: role, ScopeKind: ScopeSovereign}}
	}
	cust := func(role string) []store.RoleBinding {
		return []store.RoleBinding{{Role: role, ScopeKind: ScopeCustomer, CustomerID: ptr(a)}}
	}
	// A partner binding carries its EXPANSION: the customers of the partner
	// plus the partner's own party (DESIGN.md §Partners).
	partner := func(role, customer string) []store.RoleBinding {
		return []store.RoleBinding{{Role: role, ScopeKind: ScopePartner, PartnerID: ptr(p), Customers: []string{customer}}}
	}
	cases := []struct {
		name     string
		bindings []store.RoleBinding
		perm     Permission
		customer string
		want     bool
	}{
		{"sovereign-admin edits access", sov(RoleSovereignAdmin), SettingsManage, "", true},
		{"billing-operator cannot edit access", sov(RoleBillingOperator), SettingsManage, "", false},
		{"billing-operator issues", sov(RoleBillingOperator), BillingIssue, "", true},
		{"billing-operator manages any customer's users through customers.manage", sov(RoleBillingOperator), CustomerSelfManage, a, true},
		{"finance-viewer reads", sov(RoleFinanceViewer), MeteringRead, a, true},
		{"finance-viewer cannot issue", sov(RoleFinanceViewer), BillingIssue, "", false},
		{"finance-viewer cannot change a rate", sov(RoleFinanceViewer), RatingManage, "", false},
		{"customer-owner manages own users", cust(RoleCustomerOwner), CustomerSelfManage, a, true},
		{"customer-owner tops up", cust(RoleCustomerOwner), AccountTopup, a, true},
		{"customer-owner cannot record a payment on its own behalf", cust(RoleCustomerOwner), BillingCollect, a, false},
		{"customer-owner cannot manage customers Sovereign-wide", cust(RoleCustomerOwner), CustomersManage, a, false},
		{"customer-billing tops up", cust(RoleCustomerBilling), AccountTopup, a, true},
		{"customer-billing cannot manage users", cust(RoleCustomerBilling), CustomerSelfManage, a, false},
		{"customer-viewer reads", cust(RoleCustomerViewer), MeteringRead, a, true},
		{"customer-viewer cannot top up", cust(RoleCustomerViewer), AccountTopup, a, false},
		{"partner-owner reads its customer", partner(RolePartnerOwner, a), MeteringRead, a, true},
		{"partner-owner cannot read another partner's customer", partner(RolePartnerOwner, a), MeteringRead, b, false},
		{"partner-owner cannot manage partners Sovereign-wide", partner(RolePartnerOwner, a), PartnersManage, a, false},
		{"partner-owner cannot change a rate", partner(RolePartnerOwner, a), RatingManage, a, false},
		{"partner-owner cannot issue", partner(RolePartnerOwner, a), BillingIssue, a, false},
		{"partner-viewer reads", partner(RolePartnerViewer, a), MeteringRead, a, true},
		{"partner-viewer cannot edit the retail rule", partner(RolePartnerViewer, a), PartnerSelfManage, a, false},
		{"partner-viewer cannot top up", partner(RolePartnerViewer, a), AccountTopup, a, false},
		{"billing-operator manages any partner's rule through partners.manage", sov(RoleBillingOperator), PartnerSelfManage, a, true},
		{"no bindings hold nothing", nil, MeteringRead, a, false},
		{"an unknown role holds nothing", []store.RoleBinding{{Role: "root", ScopeKind: ScopeSovereign}}, MeteringRead, "", false},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := Has(c.bindings, c.perm, c.customer); got != c.want {
				t.Fatalf("Has(%s, %q) = %v, want %v", c.perm, c.customer, got, c.want)
			}
		})
	}
}

func TestPrimaryPicksTheHighestPowerBindingInLegacyVocabulary(t *testing.T) {
	a, b := "aaaaaaaa-0000-0000-0000-000000000001", "bbbbbbbb-0000-0000-0000-000000000002"
	cases := []struct {
		name     string
		bindings []store.RoleBinding
		role     string
		customer *string
		ok       bool
	}{
		{"none", nil, "", nil, false},
		{"viewer then owner on two customers → owner's customer", []store.RoleBinding{
			{Role: RoleCustomerViewer, ScopeKind: ScopeCustomer, CustomerID: ptr(b)},
			{Role: RoleCustomerOwner, ScopeKind: ScopeCustomer, CustomerID: ptr(a)},
		}, store.RoleCustomerAdmin, ptr(a), true},
		{"finance-viewer beats a customer-owner", []store.RoleBinding{
			{Role: RoleCustomerOwner, ScopeKind: ScopeCustomer, CustomerID: ptr(a)},
			{Role: RoleFinanceViewer, ScopeKind: ScopeSovereign},
		}, RoleFinanceViewer, nil, true},
		{"sovereign-admin reads as operator", []store.RoleBinding{{Role: RoleSovereignAdmin, ScopeKind: ScopeSovereign}}, store.RoleOperator, nil, true},
		{"customer-billing has no legacy name", []store.RoleBinding{{Role: RoleCustomerBilling, ScopeKind: ScopeCustomer, CustomerID: ptr(a)}}, RoleCustomerBilling, ptr(a), true},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			role, cust, ok := Primary(c.bindings)
			if ok != c.ok || role != c.role || !reflect.DeepEqual(cust, c.customer) {
				t.Fatalf("Primary = (%q, %v, %v), want (%q, %v, %v)", role, deref(cust), ok, c.role, deref(c.customer), c.ok)
			}
		})
	}
}

func deref(p *string) string {
	if p == nil {
		return "<nil>"
	}
	return *p
}

// A session built the old way (Role + CustomerID, no Roles) is authorised by
// the same policy, so nothing that constructs sessions that way regresses.
func TestBindingsFromLegacySession(t *testing.T) {
	a := "aaaaaaaa-0000-0000-0000-000000000001"
	op := store.Session{Role: store.RoleOperator}
	if !Has(Bindings(op), SettingsManage, "") {
		t.Fatal("legacy operator session lost settings.manage")
	}
	adm := store.Session{Role: store.RoleCustomerAdmin, CustomerID: &a}
	if !Has(Bindings(adm), CustomerSelfManage, a) || Has(Bindings(adm), CustomersManage, "") {
		t.Fatal("legacy customer-admin session maps wrong")
	}
	view := store.Session{Role: store.RoleCustomerViewer, CustomerID: &a}
	if Has(Bindings(view), AccountTopup, a) || !Has(Bindings(view), MeteringRead, a) {
		t.Fatal("legacy customer-viewer session maps wrong")
	}
	orphan := store.Session{Role: store.RoleCustomerViewer}
	if len(Bindings(orphan)) != 0 {
		t.Fatal("a customer role without a customer must bind nothing")
	}
	// Resolved Roles win over the legacy pair when both are present.
	both := store.Session{Role: store.RoleOperator, Roles: []store.RoleBinding{{Role: RoleFinanceViewer, ScopeKind: ScopeSovereign}}}
	if Has(Bindings(both), SettingsManage, "") {
		t.Fatal("resolved bindings must take precedence over the legacy role")
	}
}

func TestEffectiveAndScopes(t *testing.T) {
	a := "aaaaaaaa-0000-0000-0000-000000000001"
	bs := []store.RoleBinding{
		{Role: RoleCustomerOwner, ScopeKind: ScopeCustomer, CustomerID: ptr(a)},
		{Role: RoleFinanceViewer, ScopeKind: ScopeSovereign},
		{Role: RoleCustomerViewer, ScopeKind: ScopeCustomer, CustomerID: ptr(a)}, // duplicate coverage, must not duplicate output
	}
	eff := Effective(bs)
	if got := eff[ScopeSovereign]; !reflect.DeepEqual(got, []Permission{MeteringRead, AuditRead}) {
		t.Fatalf("sovereign permissions = %v", got)
	}
	if got := eff["customer:"+a]; !reflect.DeepEqual(got, []Permission{MeteringRead, AccountTopup, CustomerSelfManage}) {
		t.Fatalf("customer permissions = %v", got)
	}
	if got := Scopes(bs); !reflect.DeepEqual(got, []string{ScopeSovereign, "customer:" + a}) {
		t.Fatalf("scopes = %v", got)
	}
	if !IsSovereign(bs) || !OnCustomer(bs, "other") {
		t.Fatal("a Sovereign binding covers every customer")
	}
	only := []store.RoleBinding{{Role: RoleCustomerViewer, ScopeKind: ScopeCustomer, CustomerID: ptr(a)}}
	if IsSovereign(only) || OnCustomer(only, "other") || !OnCustomer(only, a) {
		t.Fatal("a customer binding covers its customer only")
	}
}

// Every role's scope kind is fixed, and the store agrees with the policy.
func TestEveryRoleHasExactlyOneScopeKind(t *testing.T) {
	for _, role := range store.Roles {
		kind := store.ScopeKindOfRole(role)
		if kind != ScopeSovereign && kind != ScopeCustomer && kind != ScopePartner {
			t.Fatalf("%s has no scope kind", role)
		}
		if _, ok := Describe[role]; !ok {
			t.Fatalf("%s has no description", role)
		}
	}
	if store.ScopeKindOfRole("root") != "" || store.ValidRole("root") {
		t.Fatal("an unknown role must be invalid")
	}
}
